test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -2702,90 +2735,117 @@ class test_ucp_sockaddr_protocols_err_sender
+        static constexpr size_t size = 64 * UCS_KBYTE;
+        std::vector<ucp_tag_message_h> messages;
+        std::vector<void*> reqs;
+        ucs_status_t status;
+
+        /* If the sumber of senders greater than 1, send the same buffer on
+         * multiple connections to delay the completion of md_invalidate on
+         * the closed connection */
+        mem_buffer send_buf(size, UCS_MEMORY_TYPE_HOST);
+        send_buf.pattern_fill(1, size);
+        for (size_t sender_idx = 0; sender_idx < num_senders; ++sender_idx) {
+            ucp_send_nbx_callback_t send_cb;
+
+            if (sender_idx > 0) {
+                send_cb = scomplete_always_ok_cbx;
+                client_ep_connect(sender_idx);
+            } else {
+                send_cb = scomplete_cbx;
+            }
+
+            /* Warmup */
+            send_recv(sender(), receiver(), send_recv_type(), false, cb_type(),
+                      sender_idx);
+
+            void *sreq = send(sender(), send_buf.ptr(), size, SEND_RECV_TAG,
+                              send_cb, NULL, sender_idx);
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
+            reqs.push_back(sreq);
+
+            /* Allow receiver to get RTS notification, but do not receive message
+             * body */
+            ucp_tag_recv_info_t info;
+            ucp_tag_message_h message = message_wait(receiver(), 0, 0, &info);
+            ASSERT_NE((void*)NULL, message);
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
+
+            messages.emplace_back(message);
+        }
+
+        /* Ignore all errors - it is expected */
+        scoped_log_handler slh(wrap_errors_logger);
+
+        /* Close the first sender's EP to force send operation to be completed
+         * with CANCEL status */
+        entity_disconnect(sender());
+
+        mem_buffer recv_buf(size, UCS_MEMORY_TYPE_HOST);
+        mem_buffer extra_recv_buf(size, UCS_MEMORY_TYPE_HOST);
+
+        recv_buf.pattern_fill(2, size);
+        extra_recv_buf.pattern_fill(2, size);
+        for (size_t i = 1; i < messages.size(); ++i) {
+            void *rreq = recv(receiver(), extra_recv_buf.ptr(), size,
+                              messages[i], rtag_complete_always_ok_cbx, NULL);
+            reqs.push_back(rreq);
+        }
+
+        status = requests_wait(reqs);
+        ASSERT_EQ(UCS_ERR_CANCELED, status);
+
+        /* Receive buffer should not be updated */
+        recv_buf.pattern_check(2, size);
+        /* Update send buffer by new data - emulation of free(buffer) */
+        send_buf.pattern_fill(3, size);
+
+        /* Complete receive operation */
+        void *rreq = recv(receiver(), recv_buf.ptr(), size, messages[0],
+                          rtag_complete_check_data_cbx, recv_buf.ptr());
+        ASSERT_TRUE(UCS_PTR_IS_PTR(rreq));
+        status = request_wait(rreq);
+        EXPECT_TRUE(UCS_STATUS_IS_ERR(status));
```

Comment:
why needed?
it seems we already have an option for success in rtag_complete_check_data_cbx
We just need to validate data in case of success, not necessarily require a failure. Seems it depends on the internal protocol.

Reply:
removed